File: CSharp/DP.UnitTest/UtVisitor.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using DP.Domain.Samples.Visitor;
using DP.UnitTest.Utility;
using Xunit;
using Xunit.Abstractions;

namespace DP.UnitTest
{
    public class UtVisitor
    {
        private readonly ITestOutputHelper _output;
        private List<IElement> _shopcart = null;

        public UtVisitor(ITestOutputHelper output)
        {
            this._output = output;
            System.Diagnostics.Trace.Listeners.Add(new XunitTraceListener(this._output));
            this._shopcart = new List<IElement>(){
                new Product { ProductType=ProductTypeEnum.Book, Name="設計模式的解析與活用", UnitPrice=480, Amount=20 },
                new Product { ProductType=ProductTypeEnum.Book, Name="使用者故事對照", UnitPrice=580, Amount=5 },
                new Product { ProductType=ProductTypeEnum.Living, Name="吸塵器", UnitPrice=2000, Amount=2 },
                new Product { ProductType=ProductTypeEnum.Living, Name="毛巾", UnitPrice=50, Amount=10 },
                new Product { ProductType=ProductTypeEnum.Electronic, Name="Surface Pro", UnitPrice=50000, Amount=2 }                
            };
        }

        [Fact]
        public void TestVisitorDiscount4Amount()
        {
            decimal expected = 
                this._shopcart[0].UnitPrice*this._shopcart[0].Amount*0.8M +
                this._shopcart[1].UnitPrice*this._shopcart[1].Amount;
            decimal actual = 0;

            IObjectStructure checkout = new ObjectStructure();

            //Attach the elements into ObjectStructure
            this._shopcart.Where(item=>item.ProductType.Equals(ProductTypeEnum.Book)).ToList().ForEach(item => {
                checkout.Attach(item);
            });
            
            //Accept all the elements and execute the strategy from certain Visitor 
            checkout.Accept(new VisitorDiscount4Count());

            actual = checkout.Elements.Sum(x=>x.TotalPrice);
            Assert.Equal(expected, actual);
        }

        [Fact]
        public void TestVisitorDiscount4TotalPrice()
        {
            decimal expected = 
                this._shopcart[2].UnitPrice*this._shopcart[2].Amount*0.9M +
                this._shopcart[3].UnitPrice*this._shopcart[3].Amount;
            decimal actual = 0;

            IObjectStructure checkout = new ObjectStructure();

            //Attach the elements into ObjectStructure
            this._shopcart.Where(item=>item.ProductType.Equals(ProductTypeEnum.Living)).ToList().ForEach(item => {
                checkout.Attach(item);
            });
            
            //Accept all the elements and execute the strategy from certain Visitor 
            checkout.Accept(new VisitorDiscount4TotalPrice());

            actual = checkout.Elements.Sum(x=>x.TotalPrice);
            Assert.Equal(expected, actual);
        }

    }
}

File: Python/Samples/Visitor/UtVisitor.py
import unittest
from decimal import Decimal,getcontext
import Elements
from Visitors import VisitorDiscount4Count, VisitorDiscount4TotalPrice
from ObjectStructure import ObjectStructure


class UtVisitor(unittest.TestCase):

    _shopcart = None

    def __init__(self, *args, **kwargs):
        super(UtVisitor, self).__init__(*args, **kwargs)
        self._shopcart = [
            Elements.Product(productType=Elements.ProductTypeEnum.Book,
                             name="設計模式的解析與活用", unitPrice=480, amount=20),
            Elements.Product(productType=Elements.ProductTypeEnum.Book,
                             name="使用者故事對照", unitPrice=580, amount=5),
            Elements.Product(productType=Elements.ProductTypeEnum.Living,
                             name="吸塵器", unitPrice=2000, amount=2),
            Elements.Product(productType=Elements.ProductTypeEnum.Living,
                             name="毛巾", unitPrice=50, amount=10),
            Elements.Product(productType=Elements.ProductTypeEnum.Electronic,
                             name="Surface Pro", unitPrice=50000, amount=2)
        ]

    def test_visitorDiscount4Amount(self):
        # getcontext().prec = 2
        
        expected = self._shopcart[0].unitPrice * Decimal(self._shopcart[0].amount) * Decimal(0.8) + self._shopcart[1].unitPrice * Decimal(self._shopcart[1].amount)
        actual = 0

        checkout = ObjectStructure()

        # Attach the elements into ObjectStructure
        targetProds = [
            x for x in self._shopcart if x.productType == Elements.ProductTypeEnum.Book]
        for item in targetProds:
            checkout.attach(item)

        # Accept all the elements and execute the strategy from certain Visitor
        checkout.accept(VisitorDiscount4Count())

        for item in checkout.elements:
            actual = actual + item.totalPrice

        self.assertEqual(actual, expected)

    def test_visitorDiscount4TotalPrice(self):
        # getcontext().prec = 2

        # expected = float(self._shopcart[2].unitPrice) * float(self._shopcart[2].amount) * float(0.9) + float(self._shopcart[3].unitPrice) * float(self._shopcart[3].amount)
        expected = self._shopcart[2].unitPrice * Decimal(self._shopcart[2].amount) * Decimal(0.9) + self._shopcart[3].unitPrice * Decimal(self._shopcart[3].amount)        
        actual = 0

        checkout = ObjectStructure()

        # Attach the elements into ObjectStructure
        targetProds = [
            x for x in self._shopcart if x.productType == Elements.ProductTypeEnum.Living]
        for item in targetProds:
            checkout.attach(item)

        # Accept all the elements and execute the strategy from certain Visitor
        checkout.accept(VisitorDiscount4TotalPrice())

        for item in checkout.elements:
            actual = actual + item.totalPrice

        self.assertEqual(actual, expected)


if __name__ == '__main__':
    unittest.main()
